ipa-core/src/protocol/context/dzkp_validator.rs

```diff
@@ -286,25 +309,27 @@ impl MultiplicationInputsBatch {
         // update last record
         self.last_record = cmp::max(self.last_record, record_id);
 
-        // panics when record_id is less than first_record
-        let id_within_batch = usize::from(record_id) - usize::from(self.first_record);
-        let block_id = (segment.len() * id_within_batch) >> BIT_ARRAY_SHIFT;
-
         // panics when record_id is too large to fit in, i.e. when it is out of bounds
-        if 256 % segment.len() == 0 {
-            self.insert_segment_small(id_within_batch, block_id, segment);
+        if segment.len() <= 256 {
+            self.insert_segment_small(record_id, segment);
         } else {
-            self.insert_segment_large(block_id, &segment);
+            self.insert_segment_large(record_id, &segment);
         }
     }
 
-    /// insert `segments` for `segments` that divide 256
+    /// insert `segments` for `segments` that are smallere than 256
     ///
     /// ## Panics
     /// Panics when `bit_length` and `block_id` are out of bounds.
-    fn insert_segment_small(&mut self, id_within_batch: usize, block_id: usize, segment: Segment) {
+    fn insert_segment_small(&mut self, record_id: RecordId, segment: Segment) {
+        // panics when record_id is less than first_record
```

Comment:
This should probably go in the panics section of the doc comment (and again, ditto for `insert_segment_large`). It might be worth doing the checked sub explicitly to make the failure cause more apparent and provide some safety in case future code changes move the indexing operation farther away from the subtraction that produces the out of bounds index.